internal/env/env.go

```diff
@@ -0,0 +1,169 @@
+package env
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func (e *sysenv) Clearenv() {
+	os.Clearenv()
+}
+
+func (e *sysenv) Setenv(k, v string) {
+	os.Setenv(k, v)
+}
+
+func SystemEnvironment() Environment {
+	return &sysenv{}
+}
+
+func NewLoader(environ ...string) *Loader {
+	if len(environ) == 0 {
+		environ = os.Environ()
+	}
+
+	var envdir string
+	original := make([]iterItem, 0, len(environ))
+	for _, v := range environ {
+		i := strings.IndexByte(v, '=')
+		if i <= 0 || i >= len(v)-1 {
+			continue
+		}
+		original = append(original, iterItem{
+			key:   v[:i],
+			value: v[i+1:],
+		})
+		if v[:i] == "ENVDIR" {
+			envdir = v[i+1:]
+		}
+	}
+
+	return &Loader{
+		original: original,
+		envdir:   envdir,
+	}
+}
+
+func (l *Loader) Restore(octx context.Context, e Environment) error {
+	return l.Apply(octx, e, WithLoadEnvdir(false))
+}
+
+func (l *Loader) Apply(octx context.Context, e Environment, options ...Option) error {
+	ctx, cancel := context.WithCancel(octx)
+	defer cancel()
+
+	e.Clearenv()
+	iter := l.Iterator(ctx, options...)
+	for iter.Next() {
+		k, v := iter.KV()
+		e.Setenv(k, v)
+	}
+
+	return nil
+}
+
+func (l *Loader) Environ(octx context.Context, options ...Option) []string {
+	ctx, cancel := context.WithCancel(octx)
+	defer cancel()
+
+	var environ []string
+	it := l.Iterator(ctx, options...)
+	for it.Next() {
+		k, v := it.KV()
+		environ = append(environ, k+`=`+v)
+	}
+	return environ
+}
+
+func (l *Loader) Iterator(ctx context.Context, options ...Option) *Iterator {
```

Comment:
What's the point of this iterator? Simply creating a slice of the environment variables would make the code much easier to follow.